core/src/main/java/net/opentsdb/query/processor/groupby/GroupByNumericArrayIterator.java

```diff
@@ -125,137 +120,131 @@ public GroupByNumericArrayIterator(final QueryNode node,
+      TemporalAmount temporalAmount = this.result.downstreamResult().timeSpecification().interval();
+      TimeStamp ts = node.pipelineContext().query().startTime().getCopy();
+      int intervals = 0;
+      while (ts.compare(Op.LT, node.pipelineContext().query().endTime())) {
+        intervals++;
+        ts.add(temporalAmount);
+      }
+      size = intervals;
+    } else {
+      size = downsampleConfig.intervals();
     }
-    
-    logger.info("Sources size {}, Intervals found {}", sources.size(), intervals);
 
-    valuesCombiner = new NumericArrayAggregator[NUM_THREADS];
-    for (int i = 0; i < valuesCombiner.length; i++) {
-      valuesCombiner[i] = factory.newAggregator(((GroupByConfig) node.config()).getInfectiousNan());
+    logger.trace("Group size: " + size);
+
+    aggregator =
+        factory.newAggregator(((GroupByConfig) node.config()).getInfectiousNan());
+    if (aggregator == null) {
+      throw new IllegalArgumentException(
+          "No aggregator found of type: " + ((GroupByConfig) node.config()).getAggregator());
     }
 
     // TODO: Need to check if it makes sense to make this threshold configurable
     if (sources.size() >= NUM_THREADS) {
+      NumericArrayAggregator[] valuesCombiner = new NumericArrayAggregator[NUM_THREADS];
+      for (int i = 0; i < valuesCombiner.length; i++) {
+        valuesCombiner[i] = createAggregator(node, factory, size);
+      }
+
+      accumulateInParallel(sources, valuesCombiner);
 
-      List<Future<TimeSeriesValue<NumericArrayType>>> futures =
-          new ArrayList<Future<TimeSeriesValue<NumericArrayType>>>(sources.size());
-      
-      parallel(sources, futures);
-//      parallelFetch(sources, futures);
-      long endTs = DateTime.nanoTime() - startTs;
-      logger.info("Total number of paralelly fetched iterator values {} and parallel fetch took {}",
-          futures.size(), ((double) endTs) / 1000_000);
-      accumulate();
     } else {
       for (final TimeSeries source : sources) {
-        if (source == null) {
-          throw new IllegalArgumentException(
-              "Null time series are not " + "allowed in the sources.");
-        }
-
-        final Optional<TypedTimeSeriesIterator<? extends TimeSeriesDataType>> optional =
-            source.iterator(NumericArrayType.TYPE);
-        if (optional.isPresent()) {
-          final TypedTimeSeriesIterator<? extends TimeSeriesDataType> iterator = optional.get();
-          if (iterator.hasNext()) {
-            has_next = true;
-            final TimeSeriesValue<NumericArrayType> array =
-                (TimeSeriesValue<NumericArrayType>) iterator.next();
-            accumulate(array);
-
-          }
-        }
+        accumulate(source, null);
       }
     }
-    
+  }
 
-    double e = ((double) (DateTime.nanoTime() - st)) / 1000_000;
-    logger.info("Groupby iterator took {}, timeseries count {}", e, sources.size());
+  private NumericArrayAggregator createAggregator(
```

Comment:
Need's `final`s and we'll need a flag to determine if we should fill and set doubles or keep longs.